src/tests/tribol_enzyme_nodal_normal.cpp

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and
+// other Tribol Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (MIT)
+
+//-----------------------------------------------------------------------------
+//
+// file: tribol_enzyme_nodal_normal.cpp
+//
+//-----------------------------------------------------------------------------
+
+#include <iostream>
+
+#include "tribol/config.hpp"
+
+#include "gtest/gtest.h"
+
+#ifdef TRIBOL_USE_UMPIRE
+#include "umpire/ResourceManager.hpp"
+#endif
+
+#include "tribol/mesh/MeshData.hpp"
+#include "tribol/geom/NodalNormal.hpp"
+#include "tribol/interface/tribol.hpp"
+
+namespace tribol {
+
+/**
+ * @brief Test fixture for the Enzyme-based derivatives of nodal normal calculations.
+ */
+class EnzymeNodalNormalTest : public testing::Test {
+ protected:
+  double delta_{ 1.0e-7 };
+  void SetUp() override {}
+
+  void FDCheck( RealT* x, MeshData& mesh )
+  {
+    VertexAvgNormal norm( true );
+    norm.Compute( mesh );
+    auto num_dofs = mesh.numberOfNodes() * mesh.spatialDimension();
+    mfem::SparseMatrix dndx( num_dofs );
+    auto& dndx_data = norm.getJacobianData();
+    // get nonmortar/nonmortar contributions
+    auto& elem_Js =
+        dndx_data.getBlockJ()( static_cast<int>( BlockSpace::NONMORTAR ), static_cast<int>( BlockSpace::NONMORTAR ) );
+    // assemble element contributions to a global sparse matrix
+    auto num_nodes_per_el = mesh.numberOfNodesPerElement();
+    for ( int i{ 0 }; i < mesh.numberOfElements(); ++i ) {
+      for ( int d1{ 0 }; d1 < mesh.spatialDimension(); ++d1 ) {
+        for ( int j{ 0 }; j < num_nodes_per_el; ++j ) {
+          auto global_idx_j = mesh.numberOfNodes() * d1 + mesh.getGlobalNodeId( i, j );
+          auto local_idx_j = num_nodes_per_el * d1 + j;
+          for ( int d2{ 0 }; d2 < mesh.spatialDimension(); ++d2 ) {
+            for ( int k{ 0 }; k < num_nodes_per_el; ++k ) {
+              auto global_idx_k = mesh.numberOfNodes() * d2 + mesh.getGlobalNodeId( i, k );
+              auto local_idx_k = num_nodes_per_el * d2 + k;
+              dndx.Add( global_idx_j, global_idx_k, elem_Js[i]( local_idx_j, local_idx_k ) );
+            }
+          }
+        }
+      }
+    }
+    dndx.Finalize();
+    mfem::DenseMatrix dndx_enzyme;
+    dndx.ToDenseMatrix( dndx_enzyme );
+
+    mfem::DenseMatrix dndx_fd( num_dofs );
+    auto mesh_view = mesh.getView();
+    norm = VertexAvgNormal( false );
```

Comment:
add comment as to what this does? Which normal calculation is then being used?

Reply:
comment added